Revert change causing regression in DAQ/HLT JSON input parsing

smorovic opened the pull request:
#### PR description:
Reverts json_reader.cpp modification which was part of this commit:
https://github.com/cms-sw/cmssw/commit/d030f663df51846c06f626488b8dfa2490d187bd

According to description, this commit fixes CLANG-10 warnings, however it seems to change behavior as well. Running with it creates this error (see below for test script description):
```
%MSG-e EvFDaqDirector:  (NoModuleName) 19-Jun-2020 20:37:32 CEST pre-events
grabNextFile Failed to deserialize JSON file -: "./data/run000100/run000100_ls0001_index000000_pid28154.jsn"
ERROR:
* Line 2, Column 20
  Missing ',' or ']' in array declaration
CONTENT:
{
   "data" : [ "20" ],
   "definition" : "./ramdisk//run000100/jsd/rawData.jsd",
   "source" : "lxplus709.cern.ch_28075"
}
.
%MSG
%MSG-e EvFDaqDirector:  (NoModuleName) 19-Jun-2020 20:37:32 CEST pre-events
grabNextFile runtime Exception -: Cannot deserialize input JSON file
%MSG
cmsRun: /tmp/smorovic/CMSSW_11_1_0_pre8/src/EventFilter/Utilities/src/FedRawDataInputSource.cc:1013: void FedRawDataInputSource::readSupervisor(): Assertion `eventsInNewFile >= 0' failed.
```
Assertion is consequence of the previous failure.

#### PR validation:
Tested manually and traced to be introduced between CMSSW_11_1_0_pre7 and CMSSW_11_1_0_pre8.
Test scripts in EventFilter/Utilities/test:
 - startBU.py (terminate manually after 10 seconds) and startFU.py, or
 - LocalRunBUFU.sh (with a log file in /tmp)

Failure mode is currently not covered with any automatic unit tests, as there have been issues implementing and running those reliably with DAQ file locking mechanism (also, for Run3 use of JSON files at HLT input stage is being phased out).

Equivalent PR will be created also for 11_1_X branch because currently this breaks release use in DAQ/HLT.

smorovic committed:
revert commit which resulted in regression in JSON parsing of DAQ input
source:
https://github.com/cms-sw/cmssw/commit/d030f663df51846c06f626488b8dfa2490d187bd

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30315/16297

- This PR adds an extra 12KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  EventFilter/Utilities/src/json_reader.cpp  modified in PR(s): #29935, #29936

cmsbuild commented:
A new Pull Request was created by  @smorovic (Srecko Morovic) for master.

It involves the following packages:

EventFilter/Utilities


@perrotta, @mommsen, @cmsbuild, @emeschi, @slava77 can you please review it and eventually sign? Thanks.
@Martin-Grunewald this is something you requested to watch as well.
@silviodonato, @dpiparo you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

smuzaffar commented on EventFilter/Utilities/src/json_reader.cpp:
```diff
@@ -430,7 +430,7 @@ namespace Json {
       }
-      bool badTokenType = (token.type_ == tokenArraySeparator || token.type_ == tokenArrayEnd);
+      bool badTokenType = (token.type_ == tokenArraySeparator && token.type_ == tokenArrayEnd);
       if (!ok || badTokenType) {
```
This does not make sense, how can this condition be true. As it will always be false so better to remove it

smorovic replied:
You're right. I bisected and found what broke the code, but didn't spot that original condition is always false.

smorovic replied:
jsoncpp seems to support comment at the end of array, which is non-standard. The check is part of this (and buggy). That feature is not relevant for what it is used in Event Filter code, so I dropped the condition with another commit.

smorovic committed:
on suggestion that original condition is always false, it can simply be
removed

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30315/16298

- This PR adds an extra 16KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  EventFilter/Utilities/src/json_reader.cpp  modified in PR(s): #29935, #29936

cmsbuild commented:
Pull request #30315 was updated. @perrotta, @mommsen, @cmsbuild, @emeschi, @slava77 can you please check and sign again.

perrotta commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: a2edc30f7722636c9e7978019454ffbde287cf97
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-95e92c/7237/summary.html
CMSSW: CMSSW_11_2_X_2020-06-21-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-95e92c/7237/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2778811
 - DQMHistoTests: Total failures: 4
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2778757
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

perrotta commented:
So, this PR reverts to the situation that was there before #29592, when ```badTokenType``` was always identically ```false```.
As long as DAQ confirms that such a ```badTokenType``` was really useless (i.e. there was not a typo in it, as #29592 implicitly assumed) this can be considered to be ok also by reco

smorovic commented:
@perrotta this code is from the external project included in the tree long time ago. It has also diverged, getting e.g. thread safety and various other fixes from CMSSW devs.

I had removed the condition because I couldn't figure out the original intention. But now I checked the original project and found out that the bug was fixes by original authors (problem was, '==' operators are supposed be '!='). Commit in their project:
https://github.com/open-source-parsers/jsoncpp/commit/9c98f2277b2972ddee93c97bb5273e86d6045195

I will test it first and then update both PRs. It is more 'proper' although either way would work fine for DAQ.

smorovic committed:
Apply jsoncpp upstream fix:
https://github.com/open-source-parsers/jsoncpp/commit/9c98f2277b2972ddee93c97bb5273e86d6045195

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30315/16324

- This PR adds an extra 16KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  EventFilter/Utilities/src/json_reader.cpp  modified in PR(s): #29935, #29936

cmsbuild commented:
Pull request #30315 was updated. @perrotta, @mommsen, @cmsbuild, @emeschi, @slava77 can you please check and sign again.

smorovic commented:
I applied the described bugfix and tested prior to pushing it to both PRs.

perrotta commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: f7d419a6ed33597ed4ad914335fe4827a21e7dc2
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-95e92c/7245/summary.html
CMSSW: CMSSW_11_2_X_2020-06-22-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

emeschi commented:
+1

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-95e92c/7245/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2778811
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2778760
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

perrotta commented:
+1
- Fixed a wrong expression implemented with #29592: original intention behind the code is now restored
- Jenkinns tests pass

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
